src/libutil/util.cc

```diff
@@ -1712,6 +1729,9 @@ void restoreMountNamespace()
     } catch (Error & e) {
         debug(e.msg());
     }
+    if (savedCwd && chdir(savedCwd->c_str()) == -1) {
```

Comment:
Rather than saving the current directory in `saveMountNamespace()`, it may be better to save a file descriptor to the current directory just before calling `setns()`, and then use `fchdir()` to restore it.

Reply:
The "just before calling `setns()`" part is tripping me up -- should I not just move `savedCwd` to an `AutoCloseFD` and modify the existing logic to reflect that, or does it _also_ need to be moved prior to the `setns()` call in `restoreMountNamespace()`? (Pardon my ignorance, but moving it into `restoreMountNamespace` just seems... wrong?)

Reply:
Switched to using an `fd` in [`e135d22` (#6348)](https://github.com/NixOS/nix/pull/6348/commits/e135d223f66f77f2b03a11b979d714e582364013).